### PR TITLE
API Key Authentication

### PR DESCRIPTION
**Pull request by steve-keep**
As a system user, I want to authenticate with Graphweaver using an API key/secret pair to make server-to-server requests.

https://app.clickup.com/t/2704425/ID-5950

### CODE REVIEW

File: src/examples/rest/database.sql [hunk 1]
```diff
@@ -39,6 +39,14 @@ CREATE TABLE credential (
   password VARCHAR(255) NOT NULL
 );
 
+CREATE TABLE api_key (
+  id INT AUTO_INCREMENT PRIMARY KEY,
+  api_key VARCHAR(255) NOT NULL UNIQUE,
+  secret VARCHAR(255) NOT NULL,
+	revoked BOOLEAN NOT NULL DEFAULT false,
```
Comment (tnodell) [inline src/examples/rest/database.sql:46]: nbd but is the indentation weird here?

File: src/examples/rest/src/backend/schema/auth/api-key/entity.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import { ApiKey as OrmApiKey } from '../../../entities';
+
+const acl: AccessControlList<ApiKey, AuthorizationContext> = {
+	DARK_SIDE: {
+		// Dark side user role can perform operations on any api keys
+		all: true,
+	},
+};
+
+@ReadOnly({ adminUI: false, backend: true })
```
Comment (tnodell) [inline src/examples/rest/src/backend/schema/auth/api-key/entity.ts:18]: Note: when I was working with Dylan yesterday, the decorator with the  `{ adminUI: false, backend: true }` options that was in his version of GW was `@ReadOnlyProperty` his no longer had `@ReadOnly`.

File: src/packages/auth/src/authentication/apollo/plugin.ts [hunk 3]
```diff
@@ -58,6 +63,7 @@ export const authApolloPlugin = (
 
 			// We may need to return a redirect to the client. If so, we'll set this variable.
 			const authHeader = request.http?.headers.get('authorization');
+			const apiKeyHeader = request.http?.headers.get('X-API-Key');
```
Comment (tnodell) [inline src/packages/auth/src/authentication/apollo/plugin.ts:66]: Is there anything special about 'X-API-Key'? What is X?

File: src/packages/auth/src/authentication/apollo/plugin.ts [hunk 4]
```diff
@@ -70,9 +76,35 @@ export const authApolloPlugin = (
...
+				const apiKey = await apiKeyDataProvider?.findOne({
+					key,
+				} as D);
+
+				if (!apiKey || !apiKey.secret) {
+					apiKeyVerificationFailedMessage = 'Bad Request: API Key Authentication Failed. (E0001)';
+				} else if (apiKey.revoked) {
+					apiKeyVerificationFailedMessage = 'Bad Request: API Key Authentication Failed. (E0002)';
+				} else if (await verifyPassword(secret, apiKey.secret)) {
+					contextValue.user = new UserProfile({
```
Comment (tnodell) [inline src/packages/auth/src/authentication/apollo/plugin.ts:96]: Why do we new up a UserProfile here? Not sure of another way to do this but every request will create a new profile in memory?

File: src/packages/auth/src/authentication/resolvers/api-key/base-resolver.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+				if (gqlEntityType.fromBackendEntity) {
+					return gqlEntityType.fromBackendEntity(apiKey);
+				}
+
+				return apiKey as any;
+			});
+		}
+	}
+
+	return ApiKeyBaseResolver;
```
Comment (tnodell) [inline src/packages/auth/src/authentication/resolvers/api-key/base-resolver.ts:153]: This is the abstract "hey you wanna make your own API key resolver you gotta follow this class"

File: src/packages/auth/src/authentication/resolvers/api-key/resolver.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	createApiKeyBaseResolver,
+} from './base-resolver';
+import { hashPassword } from '../../../utils/argon2id';
+
+export const createApiKeyResolver = <G extends WithId, D extends ApiKeyStorage>(
+	gqlEntityType: GraphqlEntityType<G, D>,
+	provider: BackendProvider<D, G>
+) => {
+	@Resolver()
+	class ApiKeyAuthResolver extends createApiKeyBaseResolver(gqlEntityType, provider) {
```
Comment (tnodell) [inline src/packages/auth/src/authentication/resolvers/api-key/resolver.ts:26]: And this is the one we're using in the rest example.

File: src/packages/auth/src/authentication/resolvers/api-key/resolver.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+			if (!item.key) throw new ValidationError('Create unsuccessful: Key not defined.');
+
+			if (!item.secret) throw new ValidationError('Create unsuccessful: Secret not defined.');
+
+			const secretHash = await hashPassword(item.secret);
+			const apiKey = await this.provider.createOne({
+				key: item.key,
+				secret: secretHash,
+				...(Object(item).hasOwnProperty('revoked') ? { revoked: item.revoked } : {}),
```
Comment (tnodell) [inline src/packages/auth/src/authentication/resolvers/api-key/resolver.ts:55]: Is this saying "Hey you're creating a key that's revoked so I'm setting it to revoked?

File: src/packages/auth/src/user-profile.ts [hunk 8]
```diff
@@ -1,5 +1,10 @@
 import { BaseDataEntity } from '@exogee/graphweaver';
 
+export enum UserProfileType {
+	SERVICE = 'service',
```
Comment (tnodell) [inline src/packages/auth/src/user-profile.ts:4]: Service is a "SystemUser", AKA a non-human user?

File: src/packages/end-to-end/src/__tests__/api/auth/api-key/api-key.test.ts [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+			key: 'test_revoked',
+			secret:
+				'$argon2id$v=19$m=65536,t=3,p=4$VCKXMZROC0Bg0Flbi9Khsg$NCOmYbM/wkuwUqB82JoOr0KzCyYsPd2WRGjy3cik0mk',
+			revoked: true,
+			roles: ['admin'],
+		});
+		if (testRevokedApiKey) database?.em.persistAndFlush(testRevokedApiKey);
+	});
+
+	test('should return a E0001 error when no system user is found.', async () => {
```
Comment (tnodell) [inline src/packages/end-to-end/src/__tests__/api/auth/api-key/api-key.test.ts:111]: I like these E000x messages.

**Review by tnodell — APPROVED** [review]
Looks good to me just a couple comments.

File: src/packages/auth/src/authentication/apollo/plugin.ts (hunk 3 above)
Comment (steve-keep) [reply src/packages/auth/src/authentication/apollo/plugin.ts:66]: `x` is used to indicate that this is not a standard HTTP header.

File: src/packages/auth/src/authentication/apollo/plugin.ts (hunk 4 above)
Comment (steve-keep) [reply src/packages/auth/src/authentication/apollo/plugin.ts:96]: yes, and that is ok as it is used by the permission system.

File: src/examples/rest/src/backend/schema/auth/api-key/entity.ts (hunk 2 above)
Comment (steve-keep) [reply src/examples/rest/src/backend/schema/auth/api-key/entity.ts:18]: I guess that was an older version?

**Pull request merged**